File: VB/CustomTokenEdit/CustomTokenEditTokenInfo.vb
Imports DevExpress.XtraEditors.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraEditors
Imports System.Drawing
Imports DevExpress.Utils.Drawing
Namespace Example.CustomTokenEdit
	Public Class CustomTokenEditTokenInfo
		Inherits TokenEditTokenInfo

		Public Sub New(ByVal token As TokenEditToken)
			MyBase.New(token)
		End Sub
		Public Sub New(ByVal token As TokenEditToken, ByVal _checkBoxBounds As Dictionary(Of TokenEditToken, Rectangle), ByVal _viewInfo As CustomTokenEditViewInfo)
			MyBase.New(token)
			checkBoxBounds = _checkBoxBounds
			viewInfo = _viewInfo
		End Sub

		Public ReadOnly Property CustomToken() As CustomTokenEditToken
			Get
				Return TryCast(MyBase.Token, CustomTokenEditToken)
			End Get
		End Property

		Private ReadOnly checkBoxBounds As Dictionary(Of TokenEditToken, Rectangle)
		Public ReadOnly Property CheckBoxGlyphBounds() As Rectangle
			Get
				Return checkBoxBounds(CustomToken)
			End Get
		End Property
		Private ReadOnly viewInfo As CustomTokenEditViewInfo

		Public ReadOnly Property Checked() As Boolean
			Get
				Return CustomToken.Checked
			End Get
		End Property
		Public Sub ChangeChecked()
			CustomToken.needRepaint = False
			CustomToken.Checked = Not CustomToken.Checked
		End Sub


        Private _checkBoxState As ObjectState
        Public Property CheckBoxState() As ObjectState
            Get
                Return _checkBoxState
            End Get
            Set(ByVal value As ObjectState)
                If _checkBoxState <> value Then
                    _checkBoxState = value
                    OnCheckBoxStateChanged()
				End If
			End Set
		End Property

		Private Sub OnCheckBoxStateChanged()
			viewInfo.InvalidateCheckBox()
		End Sub
	End Class
End Namespace

File: CS/TokenEdit_Glyph_InEndOfToken/CustomTokenEdit/CustomTokenEditTokenInfo.cs
using DevExpress.XtraEditors.ViewInfo;
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraEditors;
using System.Drawing;
using DevExpress.Utils.Drawing;
namespace TokenEdit_Glyph_InEndOfToken.CustomTokenEdit
{
    public class CustomTokenEditTokenInfo : TokenEditTokenInfo
    {
        public CustomTokenEditTokenInfo(TokenEditToken token) : base(token) { }
        public CustomTokenEditTokenInfo(TokenEditToken token, Dictionary<TokenEditToken, Rectangle> _checkBoxBounds, CustomTokenEditViewInfo _viewInfo) : base(token) 
        { 
            checkBoxBounds = _checkBoxBounds;
            viewInfo = _viewInfo;
        }

        public CustomTokenEditToken CustomToken { get { return base.Token as CustomTokenEditToken; } }

        readonly Dictionary<TokenEditToken, Rectangle> checkBoxBounds;
        public Rectangle CheckBoxGlyphBounds { get { return checkBoxBounds[CustomToken]; } }
        readonly CustomTokenEditViewInfo viewInfo;

        public bool Checked { get { return CustomToken.Checked; } }
        public void ChangeChecked()
        {
            CustomToken.needRepaint = false;
            CustomToken.Checked = !CustomToken.Checked;
        }

        ObjectState checkBoxState;
        public ObjectState CheckBoxState
        {
            get
            {
                return checkBoxState;
            }
            set
            {
                if (checkBoxState != value)
                {
                    checkBoxState = value;
                    OnCheckBoxStateChanged();
                }
            }
        }

        void OnCheckBoxStateChanged()
        {
            viewInfo.InvalidateCheckBox();
        }
    }
}
